onnxruntime/test/python/onnxruntime_test_python_nested_control_flow_op.py

```diff
@@ -0,0 +1,247 @@
+    )
+    expand_shape = helper.make_tensor(f"{add_const_out_vi}.shape", TensorProto.INT64, [2], [1, 2])
+    expand_node = helper.make_node(
+        "Expand",
+        inputs=[add_const_out_vi.name, expand_shape.name],
+        outputs=[y3_vi.name],
+        name="Expand.out",
+    )
+    add_x3_const_subgraph = helper.make_graph(
+        [const_node, add_const_node, expand_node],
+        name="add-x3-const-subgraph",
+        inputs=[],
+        outputs=[y3_vi],
+        value_info=[x3_vi, const_vi, add_const_out_vi],
+        initializer=[const_shape_tensor, expand_shape],
+    )
+
+    # Subgraph flow based on `x3` value
+    y3_if_vi = make_vi_like(y_vi, "x3.if.out")
+    x3_eq_vi, x3_const_node, x3_const_shape_tensor = make_const(x3_vi, "x3.equal", value=0)
+    x3_ge_vi = helper.make_tensor_value_info(
+        "x3_ge",
+        TensorProto.BOOL,
+        shape=[1],
+    )
+    x3_ge_node = helper.make_node(
+        "GreaterOrEqual",
+        inputs=[x3_vi.name, x3_eq_vi.name],
+        outputs=[x3_ge_vi.name],
+        name="GreaterOrEqual.Target",
+    )
+    x3_has_elem_vi = helper.make_tensor_value_info(
+        "x3_has_elem",
+        TensorProto.BOOL,
+        shape=[],  # scalar
+    )
+    x3_has_elem_node = helper.make_node(
+        "Squeeze",
+        inputs=[x3_ge_vi.name],
+        outputs=[x3_has_elem_vi.name],
+        name="Squeeze.x3",
+    )
+    if_input_node = helper.make_node(
+        "If",
+        inputs=[x3_has_elem_vi.name],  # condition
+        outputs=[y3_if_vi.name],
+        name="If.OptionalHasElement.x3",
+        then_branch=add_x3_const_subgraph,
+        else_branch=add_x2_const_subgraph,
+    )
+    x3_subgraph = helper.make_graph(
+        [x3_const_node, x3_ge_node, x3_has_elem_node, if_input_node],
+        name="x3-subgraph",
+        inputs=[],
+        outputs=[y3_if_vi],
+        value_info=[x3_vi, x3_eq_vi, x3_ge_vi, x3_has_elem_vi],
+        initializer=[x3_const_shape_tensor],
+    )
+
+    # Construct main graph
+    x1_has_elem_vi = helper.make_tensor_value_info(
+        "x1_has_elem",
+        TensorProto.BOOL,
+        shape=[],  # scalar
+    )
+    x1_has_elem_node = helper.make_node(
+        "OptionalHasElement",
+        inputs=[opt_x1_vi.name],
+        outputs=[x1_has_elem_vi.name],
+        name="OptionalHasElement.x1",
+    )
+    if_input_node = helper.make_node(
+        "If",
+        inputs=[x1_has_elem_vi.name],  # condition
+        outputs=[y_vi.name],
+        name="If.OptionalHasElement.x1",
+        then_branch=add_x1_x2_subgraph,
+        else_branch=x3_subgraph,
+    )
+    graph = helper.make_graph(
+        [x1_has_elem_node, if_input_node],
+        "opt-graph",
+        [opt_x1_vi, x2_vi, x3_vi],
+        [y_vi],
+        value_info=[x1_has_elem_vi],
+    )
+
+    m = helper.make_model(
+        graph,
+        opset_imports=[
+            helper.make_opsetid("", 15),
+        ],
+    )
+
+    checker.check_model(m, full_check=True)
+
+    return m
```

Comment:
I would consider a simple text representation of the model for readability and simplicity:


```python
model_text = """\
<
   ir_version: 9,
   opset_import: ["" : 15]
>
opt_graph (optional(float[1,2]) x1_opt, float[1,2] x2, float[1] x3) => (float[1,2] y) {
   x1_has_elem = OptionalHasElement (x1_opt)
   y = If (x1_has_elem) <
      else_branch = x3_subgraph () => (float[1,2] x3_if_out) {
         x3_equal = ConstantOfShape <value = float[1] x3_equal_const_value {0}> (x3_equal_shape)
         x3_ge = GreaterOrEqual (x3, x3_equal)
         x3_has_elem = Squeeze (x3_ge)
         x3_if_out = If (x3_has_elem) <
            else_branch = add_x2_const_subgraph () => (float[1,2] x2_add_const) {
               x1_const = ConstantOfShape <value = float[1] x1_const_const_value {1}> (x1_const_shape)
               x2_add_const = Add (x1_const, x2)
            }, then_branch = add_x3_const_subgraph () => (float[1,2] x3_add_const) {
               x3_const = ConstantOfShape <value = float[1] x3_const_const_value {2}> (x3_const_shape)
               out_1 = Add (x3_const, x3)
               x3_add_const = Expand (out_1, x3_const)
            }
         >
     }, then_branch = add_x1_x2_subgraph () => (float[1,2] x1_add_x2) {
        x1 = OptionalGetElement (x1_opt)
        x1_add_x2 = Add (x1, x2)
     }>
}"""

model = onnx.parser.parse_model(model_text)
```

Reply:
So no need to build the proto procedurally

Reply:
this is a good suggestion. can you implement this @chilo-ms ?

Reply:
Could you help point out how to add initializers to the model_text since initializers are missing in the text representation below? Thanks. 

> I would consider a simple text representation of the model for readability and simplicity:
> 
> ```python
> model_text = """\
> <
>    ir_version: 9,
>    opset_import: ["" : 15]
> >
> opt_graph (optional(float[1,2]) x1_opt, float[1,2] x2, float[1] x3) => (float[1,2] y) {
>    x1_has_elem = OptionalHasElement (x1_opt)
>    y = If (x1_has_elem) <
>       else_branch = x3_subgraph () => (float[1,2] x3_if_out) {
>          x3_equal = ConstantOfShape <value = float[1] x3_equal_const_value {0}> (x3_equal_shape)
>          x3_ge = GreaterOrEqual (x3, x3_equal)
>          x3_has_elem = Squeeze (x3_ge)
>          x3_if_out = If (x3_has_elem) <
>             else_branch = add_x2_const_subgraph () => (float[1,2] x2_add_const) {
>                x1_const = ConstantOfShape <value = float[1] x1_const_const_value {1}> (x1_const_shape)
>                x2_add_const = Add (x1_const, x2)
>             }, then_branch = add_x3_const_subgraph () => (float[1,2] x3_add_const) {
>                x3_const = ConstantOfShape <value = float[1] x3_const_const_value {2}> (x3_const_shape)
>                out_1 = Add (x3_const, x3)
>                x3_add_const = Expand (out_1, x3_const)
>             }
>          >
>      }, then_branch = add_x1_x2_subgraph () => (float[1,2] x1_add_x2) {
>         x1 = OptionalGetElement (x1_opt)
>         x1_add_x2 = Add (x1, x2)
>      }>
> }"""
> 
> model = onnx.parser.parse_model(model_text)
> ```

Reply:
> I would consider a simple text representation of the model for readability and simplicity:
> 
> ```python
> model_text = """\
> <
>    ir_version: 9,
>    opset_import: ["" : 15]
> >
> opt_graph (optional(float[1,2]) x1_opt, float[1,2] x2, float[1] x3) => (float[1,2] y) {
>    x1_has_elem = OptionalHasElement (x1_opt)
>    y = If (x1_has_elem) <
>       else_branch = x3_subgraph () => (float[1,2] x3_if_out) {
>          x3_equal = ConstantOfShape <value = float[1] x3_equal_const_value {0}> (x3_equal_shape)
>          x3_ge = GreaterOrEqual (x3, x3_equal)
>          x3_has_elem = Squeeze (x3_ge)
>          x3_if_out = If (x3_has_elem) <
>             else_branch = add_x2_const_subgraph () => (float[1,2] x2_add_const) {
>                x1_const = ConstantOfShape <value = float[1] x1_const_const_value {1}> (x1_const_shape)
>                x2_add_const = Add (x1_const, x2)
>             }, then_branch = add_x3_const_subgraph () => (float[1,2] x3_add_const) {
>                x3_const = ConstantOfShape <value = float[1] x3_const_const_value {2}> (x3_const_shape)
>                out_1 = Add (x3_const, x3)
>                x3_add_const = Expand (out_1, x3_const)
>             }
>          >
>      }, then_branch = add_x1_x2_subgraph () => (float[1,2] x1_add_x2) {
>         x1 = OptionalGetElement (x1_opt)
>         x1_add_x2 = Add (x1, x2)
>      }>
> }"""
> 
> model = onnx.parser.parse_model(model_text)
> ```

Could you help point it out how to add initializers to the model_text? since the text representation is missing some initializers here.

Reply:
I don't think initializers are supported right now but it should be possible to add constants with int/float values. What kind of initializers do you need?

Reply:
The `x1_const_shape` and `c3_const_shape` are the constant tensors and will be included in subgraph's initializers. But in the textual syntax `model_text` we are missing the definitions of those constant tensors so it will fail when ORT validating the graph. I'm not familiar with textual syntax, any suggestions here? thanks.